### ISSUE TITLE
@base treats trailing hash ('#') different from slash '/'

### ISSUE THREAD

**Issue by simontaurus** [json-ld/json-ld.org#809]
```jsonld
{
  "@context": {
    "@version": 1.1,
    "@base":"http://example.com/vocab/",
    "name": {"@id": "http://example.com/vocab#test", "@type": "@id"}
  },
  "@id": "http://example.org/places#BrewEats",
  "@type": "Restaurant",
  "name": "Restaurant"
}
```
expands to
```jsonld
[
  {
    "@id": "http://example.org/places#BrewEats",
    "@type": [
      "http://example.com/vocab/Restaurant"
    ],
    "http://example.com/vocab#test": [
      {
        "@id": "http://example.com/vocab/Restaurant"
      }
    ]
  }
]
```
([playground](https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22%40base%22%3A%22http%3A%2F%2Fexample.com%2Fvocab%2F%22%2C%22name%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.com%2Fvocab%23test%22%2C%22%40type%22%3A%22%40id%22%7D%7D%2C%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Fplaces%23BrewEats%22%2C%22%40type%22%3A%22Restaurant%22%2C%22name%22%3A%22Restaurant%22%7D))
while 
```jsonld
{
  "@context": {
    "@version": 1.1,
    "@base":"http://example.com/vocab#",
    "name": {"@id": "http://example.com/vocab#test", "@type": "@id"}
  },
  "@id": "http://example.org/places#BrewEats",
  "@type": "Restaurant",
  "name": "Restaurant"
}
```
ignores the hash and falls back to the last slash
```jsonld
[
  {
    "@id": "http://example.org/places#BrewEats",
    "@type": [
      "http://example.com/Restaurant"
    ],
    "http://example.com/vocab#test": [
      {
        "@id": "http://example.com/Restaurant"
      }
    ]
  }
]
```
([playground](https://json-ld.org/playground/#startTab=tab-expanded&json-ld=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22%40base%22%3A%22http%3A%2F%2Fexample.com%2Fvocab%23%22%2C%22name%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.com%2Fvocab%23test%22%2C%22%40type%22%3A%22%40id%22%7D%7D%2C%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Fplaces%23BrewEats%22%2C%22%40type%22%3A%22Restaurant%22%2C%22name%22%3A%22Restaurant%22%7D))

Is this behaviour intented? I did not found any hint in the spec

**Comment by gkellogg** [comment]
This follows [RFC3987 resolution of IRI references](https://datatracker.ietf.org/doc/html/rfc3987#page-32), which does not have special provision for IRIs ending with `#`. See the [IRI Expansion Algorithm](https://www.w3.org/TR/json-ld11-api/#algorithm-4) step 8.

> Otherwise, if document relative is true set value to the result of resolving value against the [base IRI](https://www.w3.org/TR/json-ld11-api/#dfn-context-base-iri) from active context. Only the basic algorithm in [section 5.2](https://tools.ietf.org/html/rfc3986#section-5.2) of [[RFC3986](https://www.w3.org/TR/json-ld11-api/#bib-rfc3986)] is used; neither [Syntax-Based Normalization](https://tools.ietf.org/html/rfc3986#section-6.2.2) nor [Scheme-Based Normalization](https://tools.ietf.org/html/rfc3986#section-6.2.3) are performed. ...

**Comment by simontaurus** [comment]
Thank you for the clarification. So there is no way to use a vocabulary that uses `#` as part of its IRIs as a `@base`, e. g. to expand 
```json
{"type": "string"}
```
to
```json
{"type": "http://www.w3.org/2001/XMLSchema#string"}
```
?

**Comment by gkellogg** [comment]
Certainly, you can use `@vocab` for thins like `@type` IRI expansion. `@type` is special as it can expand either relative to the vocabulary or document base. Vocabulary expansion, also used for keys, uses string concatenation, rather than IRI resolution, for this purpose.

`@base` is used to override the document location, for entities described within that document.
